File: src/main/java/com/canefe/story/LocationManager.java
package com.canefe.story;

import org.bukkit.Location;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class LocationManager {

    private static LocationManager instance; // Singleton instance
    private final Story plugin;
    private final Map<String, StoryLocation> locations;
    private final File locationDirectory;

    private LocationManager(Story plugin) {
        this.plugin = plugin;
        this.locationDirectory = new File(plugin.getDataFolder(), "locations");
        this.locations = new HashMap<>();

        if (!locationDirectory.exists()) {
            locationDirectory.mkdirs(); // Create the directory if it doesn't exist
        }
    }

    // Get the singleton instance
    public static synchronized LocationManager getInstance(Story plugin) {
        if (instance == null) {
            instance = new LocationManager(plugin);
        }
        return instance;
    }

    public File getLocationDirectory() {
        return locationDirectory;
    }

    public void saveLocationFile(String locationName, FileConfiguration config) {
        File locationFile = new File(locationDirectory, locationName + ".yml");
        try {
            config.save(locationFile);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public StoryLocation loadLocationData(String locationName) {
        File locationFile = new File(getLocationDirectory(), locationName + ".yml");
        if (!locationFile.exists()) {
            return null; // Return null if no file exists
        }
        FileConfiguration config = YamlConfiguration.loadConfiguration(locationFile);
        StoryLocation storyLocation = new StoryLocation(locationName, config.getStringList("npcNames"), config.getStringList("context"));

        locations.put(locationName, storyLocation);

        return storyLocation;
    }

    // Add a location
    public void addLocation(StoryLocation storyLocation) {
        locations.put(storyLocation.getName(), storyLocation);
    }

    // Get a location by name
    public StoryLocation getLocation(String name) {
        // Check if it's already loaded in memory first
        StoryLocation location = locations.get(name);
        if (location == null) {
            // If not loaded, load it with the method that handles Bukkit locations
            location = loadLocation(name);
            if (location != null) {
                // Store in cache for future use
                locations.put(name, location);
            }
        }
        return location;
    }

    // Get all locations
    public List<StoryLocation> getAllLocations() {
        return List.copyOf(locations.values());
    }

    // Get location-specific global contexts
    // Replace your existing getLocationGlobalContexts method
    public List<String> getLocationGlobalContexts(String locationName) {
        return getAllContextForLocation(locationName);
    }

    // Check if an NPC is part of a location
    public boolean isNPCInLocation(String npcName, String locationName) {
        StoryLocation storyLocation = locations.get(locationName);
        return storyLocation != null && storyLocation.getNpcNames().contains(npcName);
    }

// In your LocationManager class:

    public void loadAllLocations() {
        locations.clear();
        loadLocationsRecursively(locationDirectory, null);
        plugin.getLogger().info("Loaded " + locations.size() + " locations");
    }

    private void loadLocationsRecursively(File directory, String parentPath) {
        if (!directory.exists() || !directory.isDirectory()) return;

        File[] files = directory.listFiles();
        if (files == null) return;

        // First pass: load all YAML files in this directory
        for (File file : files) {
            if (file.isFile() && file.getName().endsWith(".yml")) {
                String locationName = file.getName().replace(".yml", "");
                String fullPath = parentPath == null ? locationName : parentPath + "/" + locationName;

                FileConfiguration config = YamlConfiguration.loadConfiguration(file);
                List<String> npcNames = config.getStringList("npcs");
                List<String> context = config.getStringList("context");

                // Use explicit parent from file if available, otherwise use directory structure
                String explicitParent = config.getString("parent", null);
                String effectiveParent = explicitParent != null ? explicitParent : parentPath;

                StoryLocation location = new StoryLocation(fullPath, npcNames, context, effectiveParent);

                // Load Bukkit location if exists
                if (config.contains("world")) {
                    String worldName = config.getString("world");
                    double x = config.getDouble("x");
                    double y = config.getDouble("y");
                    double z = config.getDouble("z");
                    float yaw = (float) config.getDouble("yaw");
                    float pitch = (float) config.getDouble("pitch");

                    org.bukkit.World world = plugin.getServer().getWorld(worldName);
                    if (world != null) {
                        Location bukkitLocation = new Location(world, x, y, z, yaw, pitch);
                        location.setBukkitLocation(bukkitLocation);
                    }
                }

                locations.put(fullPath, location);
                plugin.getLogger().info("Loaded location: " + fullPath +
                        (effectiveParent != null ? " with parent: " + effectiveParent : ""));
            }
        }

        // Second pass: process subdirectories
        for (File file : files) {
            if (file.isDirectory()) {
                String dirName = file.getName();
                String newParentPath = parentPath == null ? dirName : parentPath + "/" + dirName;
                loadLocationsRecursively(file, newParentPath);
            }
        }
    }

    // Updated to work with the recursive loading system
    public StoryLocation loadLocation(String name) {
        // Check if already loaded
        StoryLocation location = locations.get(name);
        if (location != null) {
            return location;
        }

        // If not loaded, try to find file - first as direct path
        File locationFile = new File(locationDirectory, name + ".yml");

        // If file doesn't exist, try with folder structure
        if (!locationFile.exists() && name.contains("/")) {
            locationFile = new File(locationDirectory, name.replace("/", File.separator) + ".yml");
        }

        if (!locationFile.exists()) {
            return null;
        }

        // Load it and add to cache
        FileConfiguration config = YamlConfiguration.loadConfiguration(locationFile);
        List<String> npcNames = config.getStringList("npcs");
        List<String> context = config.getStringList("context");
        String parentName = config.getString("parent", null);

        // If no explicit parent and this is a path with slashes, infer parent from path
        if (parentName == null && name.contains("/")) {
            parentName = name.substring(0, name.lastIndexOf("/"));
        }

        location = new StoryLocation(name, npcNames, context, parentName);

        // Load Bukkit location
        if (config.contains("world")) {
            String worldName = config.getString("world");
            double x = config.getDouble("x");
            double y = config.getDouble("y");
            double z = config.getDouble("z");
            float yaw = (float) config.getDouble("yaw");
            float pitch = (float) config.getDouble("pitch");

            org.bukkit.World world = plugin.getServer().getWorld(worldName);
            if (world != null) {
                Location bukkitLocation = new Location(world, x, y, z, yaw, pitch);
                location.setBukkitLocation(bukkitLocation);
            }
        }

        locations.put(name, location);
        return location;
    }

    public void saveLocation(StoryLocation location) {
        File locationFile = new File(locationDirectory, location.getName() + ".yml");
        FileConfiguration config = new YamlConfiguration();

        // Save basic properties
        config.set("name", location.getName());
        config.set("npcs", location.getNpcNames());
        config.set("context", location.getContext());

        // Save parent location if it exists
        if (location.hasParent()) {
            config.set("parent", location.getParentLocationName());
        }

        // Save Bukkit location if it exists
        Location bukkitLoc = location.getBukkitLocation();
        if (bukkitLoc != null) {
            config.set("world", bukkitLoc.getWorld().getName());
            config.set("x", bukkitLoc.getX());
            config.set("y", bukkitLoc.getY());
            config.set("z", bukkitLoc.getZ());
            config.set("yaw", bukkitLoc.getYaw());
            config.set("pitch", bukkitLoc.getPitch());
        }

        try {
            config.save(locationFile);
        } catch (IOException e) {
            plugin.getLogger().severe("Could not save location: " + location.getName());
            e.printStackTrace();
        }
    }

    // New method to get ALL context for a location (including parent contexts)
    public List<String> getAllContextForLocation(String locationName) {
        List<String> allContext = new ArrayList<>();
        StoryLocation location = getLocation(locationName);

        if (location == null) {
            return allContext;
        }

        // Add this location's context
        allContext.addAll(location.getContext());

        // Recursively add parent contexts
        String parentName = location.getParentLocationName();
        while (parentName != null && !parentName.isEmpty()) {
            StoryLocation parentLocation = getLocation(parentName);
            if (parentLocation == null) {
                break; // Parent not found
            }

            allContext.addAll(parentLocation.getContext());
            parentName = parentLocation.getParentLocationName();
        }

        return allContext;
    }
}

File: src/main/kotlin/com/canefe/story/location/LocationManager.kt
package com.canefe.story.location

import com.canefe.story.Story
import com.canefe.story.location.data.StoryLocation
import com.canefe.story.storage.LocationDocument
import com.canefe.story.storage.LocationStorage
import org.bukkit.Location
import java.io.File

class LocationManager(
    private val plugin: Story,
    private var locationStorage: LocationStorage,
) {
    fun updateStorage(storage: LocationStorage) {
        locationStorage = storage
    }

    private val locations: MutableMap<String, StoryLocation> = HashMap()
    val locationDirectory: File =
        File(plugin.dataFolder, "locations").apply {
            if (!exists()) mkdirs()
        }

    fun getSublocations(parentLocationName: String): List<StoryLocation> =
        locations.values.filter { location ->
            location.parentLocationName == parentLocationName ||
                location.name.startsWith("$parentLocationName/")
        }

    fun loadLocationData(locationName: String): StoryLocation? {
        val doc = locationStorage.loadLocation(locationName) ?: return null
        val storyLocation = documentToStoryLocation(doc)
        locations[locationName] = storyLocation
        return storyLocation
    }

    fun getLocationByPosition(
        position: Location,
        range: Double = 50.0,
    ): StoryLocation? {
        var closestLocation: StoryLocation? = null
        var closestDistance = Double.MAX_VALUE

        for (location in locations.values) {
            val bukkitLocation = location.bukkitLocation ?: continue
            if (bukkitLocation.world != position.world) continue

            val distance = bukkitLocation.distance(position)
            if (distance <= range && distance < closestDistance) {
                closestDistance = distance
                closestLocation = location
            }
        }

        return closestLocation
    }

    fun getLocationByPosition2D(
        position: Location,
        range: Double = 50.0,
    ): StoryLocation? {
        var closestLocation: StoryLocation? = null
        var closestDistance = Double.MAX_VALUE

        for (location in locations.values) {
            val loc = location.bukkitLocation ?: continue
            if (loc.world != position.world) continue

            val dx = loc.x - position.x
            val dz = loc.z - position.z
            val distSq = dx * dx + dz * dz

            if (distSq <= range * range && distSq < closestDistance) {
                closestDistance = distSq
                closestLocation = location
            }
        }

        return closestLocation
    }

    fun addLocation(storyLocation: StoryLocation) {
        locations[storyLocation.name] = storyLocation
    }

    fun createLocation(
        name: String,
        bukkitLocation: Location?,
    ): StoryLocation? {
        return try {
            if (locations.containsKey(name)) {
                plugin.logger.warning("Location with name $name already exists.")
                return null
            }

            val location = StoryLocation(name, "", bukkitLocation, null)
            locations[name] = location
            saveLocation(location)
            location
        } catch (e: Exception) {
            plugin.logger.severe("Failed to create location: $name")
            e.printStackTrace()
            null
        }
    }

    fun getOrCreateDefaultLocation(): StoryLocation {
        val name = plugin.configService.defaultLocationName
        return getLocation(name) ?: createLocation(name, null)
            ?: StoryLocation(name, "", null, null)
    }

    fun getLocation(name: String): StoryLocation? =
        locations[name] ?: loadLocation(name)?.also { location ->
            locations[name] = location
        }

    fun getAllLocations(): List<StoryLocation> = locations.values.toList()

    fun getLocationGlobalContexts(locationName: String): String = getAllContextForLocation(locationName)

    fun loadAllLocations() {
        locations.clear()
        val docs = locationStorage.loadAllLocations()
        for ((name, doc) in docs) {
            locations[name] = documentToStoryLocation(doc)
        }
        plugin.logger.info("Loaded ${locations.size} locations")
    }

    fun loadLocation(name: String): StoryLocation? {
        locations[name]?.let { return it }

        val doc = locationStorage.loadLocation(name) ?: return null
        val location = documentToStoryLocation(doc)
        locations[name] = location
        return location
    }

    fun saveLocation(location: StoryLocation) {
        val doc = storyLocationToDocument(location)
        locationStorage.saveLocation(doc)

        // Reload to update cache
        val reloaded = locationStorage.loadLocation(location.name)
        if (reloaded != null) {
            locations[location.name] = documentToStoryLocation(reloaded)
        }
    }

    fun saveLocationFile(
        locationName: String,
        config: org.bukkit.configuration.file.FileConfiguration,
    ) {
        val locationFile = File(locationDirectory, "$locationName.yml")
        try {
            config.save(locationFile)
        } catch (e: java.io.IOException) {
            plugin.logger.severe("Failed to save location file: $locationName")
        }
    }

    fun getAllContextForLocation(locationName: String): String {
        val parts = mutableListOf<String>()
        val location = getLocation(locationName) ?: return ""

        if (location.description.isNotBlank()) {
            parts.add(location.description)
        }

        var parentName = location.parentLocationName
        while (!parentName.isNullOrEmpty()) {
            val parentLocation = getLocation(parentName) ?: break
            if (parentLocation.description.isNotBlank()) {
                parts.add(parentLocation.description)
            }
            parentName = parentLocation.parentLocationName
        }

        return parts.joinToString("\n")
    }

    private fun documentToStoryLocation(doc: LocationDocument): StoryLocation {
        val location =
            StoryLocation(
                doc.name,
                doc.description,
                doc.parentLocationName,
            )

        if (doc.allowedNPCs.isNotEmpty()) {
            location.allowedNPCs.addAll(doc.allowedNPCs)
        }

        location.randomPathingAction = doc.randomPathingAction
        location.hideTitle = doc.hideTitle

        // Reconstruct Bukkit Location
        if (doc.world != null) {
            val world = plugin.server.getWorld(doc.world)
            if (world != null) {
                location.bukkitLocation = Location(world, doc.x, doc.y, doc.z, doc.yaw, doc.pitch)
            }
        }

        return location
    }

    private fun storyLocationToDocument(location: StoryLocation): LocationDocument {
        val bukkitLoc = location.bukkitLocation
        return LocationDocument(
            name = location.name,
            description = location.description,
            parentLocationName = if (location.hasParent()) location.parentLocationName else null,
            world = bukkitLoc?.world?.name,
            x = bukkitLoc?.x ?: 0.0,
            y = bukkitLoc?.y ?: 0.0,
            z = bukkitLoc?.z ?: 0.0,
            yaw = bukkitLoc?.yaw ?: 0f,
            pitch = bukkitLoc?.pitch ?: 0f,
            allowedNPCs = location.allowedNPCs,
            hideTitle = location.hideTitle,
            randomPathingAction = location.randomPathingAction,
        )
    }
}
